File: DP/src/main/java/Solution322.java

// https://leetcode-cn.com/problems/coin-change/
import java.util.*;

public class Solution322 {
    public int coinChange(int[] coins, int amount) {
        int[] opt = new int[amount + 1];
        Arrays.fill(opt, amount + 1);
        opt[0] = 0;
        for (int i = 1; i < amount + 1; i++) {
            for (int coin : coins) {
                if (i - coin < 0) {
                    continue;
                }

                opt[i] = Math.min(opt[i], opt[i - coin] + 1);

            }

        }

        return opt[amount] == amount + 1 ? -1 : opt[amount];
    }
}

File: DP/src/main/python/solution322.py
# https://leetcode-cn.com/problems/coin-change/
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:

        ans = float("INF")

        opt = [ans for _ in range(amount + 1)]
        opt[0] = 0

        for i in range(1, amount + 1):

            for coin in coins:

                if i - coin < 0:
                    continue

                opt[i] = min(opt[i], opt[i - coin] + 1)

        return opt[amount] if opt[amount] != float("INF") else -1
